### PR TITLE
Add check to prevent leafnode connecting to client port

### PR DESCRIPTION
**Pull request by wallyqs**
Currently when a NATS Server connects to a leafnode to the client port, the connection is established successfully and remains connected until the first message is sent.

For example given this config:

```hcl
leafnodes {
  remotes = [
    {
      url = "nats://connect.ngs.global:4222"
      credentials = "~/.nkeys/creds/synadia/NGS/NGS.creds"
    }
  ]
}
```

The server will remain connected, with ping/pong working ok:

```
2020/01/28 12:52:16.468011 [INF] Starting nats-server version 2.1.3-RC05
2020/01/28 12:52:16.468129 [DBG] Go build version go1.13.3
2020/01/28 12:52:16.468134 [INF] Git commit [not set]
2020/01/28 12:52:16.468479 [INF] Listening for client connections on 0.0.0.0:4222
2020/01/28 12:52:16.468496 [INF] Server id is NC2IUWWQCR7OYW6ONQAFQINLKCDWJ6BXGAL52L5JLDZHWJGVIMFXMA72
2020/01/28 12:52:16.468503 [INF] Server is ready
2020/01/28 12:52:16.468517 [DBG] Get non local IPs for "0.0.0.0"
2020/01/28 12:52:16.468840 [DBG]  ip=192.168.1.7
2020/01/28 12:52:16.470977 [DBG] Trying to connect as leafnode to remote server on "connect.ngs.global:4222" (35.166.100.73:4222)
2020/01/28 12:52:16.546743 [DBG] 35.166.100.73:4222 - lid:1 - Starting TLS leafnode client handshake
2020/01/28 12:52:16.791221 [DBG] 35.166.100.73:4222 - lid:1 - Authenticating with credentials file "/Users/wallyqs/.nkeys/creds/synadia/NGS/NGS.creds"
2020/01/28 12:52:16.792689 [DBG] 35.166.100.73:4222 - lid:1 - Remote leafnode connect msg sent
2020/01/28 12:52:16.792746 [DBG] 35.166.100.73:4222 - lid:1 - Leafnode connection created
2020/01/28 12:52:16.792860 [INF] Connected leafnode to "connect.ngs.global"
2020/01/28 12:52:17.847924 [DBG] 35.166.100.73:4222 - lid:1 - LeafNode Ping Timer
2020/01/28 12:52:17.849605 [TRC] 35.166.100.73:4222 - lid:1 - ->> [PING]
2020/01/28 12:52:17.909107 [TRC] 35.166.100.73:4222 - lid:1 - <<- [PONG]
2020/01/28 12:52:18.950025 [TRC] 35.166.100.73:4222 - lid:1 - <<- [PING]
2020/01/28 12:52:18.950061 [TRC] 35.166.100.73:4222 - lid:1 - ->> [PONG]
```

The connection will be in the established state until the first message is sent on the leafnode connection, after which the server will fail to parse it with an `Unknown Protocol Error`.

```
2020/01/28 12:54:43.292346 [DBG] Trying to connect as leafnode to remote server on "connect.ngs.global:4222" (35.166.100.73:4222)
2020/01/28 12:54:43.368876 [DBG] 35.166.100.73:4222 - lid:4 - Starting TLS leafnode client handshake
2020/01/28 12:54:43.464516 [DBG] 35.166.100.73:4222 - lid:4 - Authenticating with credentials file "/Users/wallyqs/.nkeys/creds/synadia/NGS/NGS.creds"
2020/01/28 12:54:43.466691 [DBG] 35.166.100.73:4222 - lid:4 - Remote leafnode connect msg sent
2020/01/28 12:54:43.466830 [DBG] 35.166.100.73:4222 - lid:4 - Leafnode connection created
2020/01/28 12:54:43.466920 [TRC] 35.166.100.73:4222 - lid:4 - ->> [LS+ _INBOX.Nn5na6z6dFjbXIbwiLmQQm.*]
2020/01/28 12:54:43.502832 [ERR] 35.166.100.73:4222 - lid:4 - Leafnode Error 'Unknown Protocol Operation'
2020/01/28 12:54:43.503007 [DBG] 35.166.100.73:4222 - lid:4 - LeafNode connection closed
```

The INFO protocol from both client and leafnode port is the same other, with the only difference so far being that the client port uses `connect_urls` for gossiping whereas the leafnode port uses
`leafnode_urls`.  In this commit we use that as the check to prevent leafnode connections using the wrong port and report with debug level any errors while connecting:

```sh
2020/01/28 12:51:48.104976 [INF] Starting nats-server version 2.1.3-RC05
2020/01/28 12:51:48.105102 [DBG] Go build version go1.13.3
2020/01/28 12:51:48.105107 [INF] Git commit [not set]
2020/01/28 12:51:48.105465 [INF] Listening for client connections on 0.0.0.0:4222
2020/01/28 12:51:48.105477 [INF] Server id is NAQ7IMSSPVYMBY5GJTE3PWVDL3ODSAC6WJ6BUM4G5VB3T4ILEU4S7BOT
2020/01/28 12:51:48.105482 [INF] Server is ready
2020/01/28 12:51:48.105493 [DBG] Get non local IPs for "0.0.0.0"
2020/01/28 12:51:48.105786 [DBG]  ip=192.168.1.7
2020/01/28 12:51:48.127007 [DBG] Trying to connect as leafnode to remote server on "connect.ngs.global:4222" (35.166.100.73:4222)
2020/01/28 12:51:48.199762 [DBG] 35.166.100.73:4222 - lid:1 - Error reading remote leafnode's INFO: attempted to connect to client port
2020/01/28 12:51:48.200044 [DBG] 35.166.100.73:4222 - lid:1 - LeafNode connection closed
2020/01/28 12:51:48.200283 [INF] Connected leafnode to "connect.ngs.global"
```

 - [X] Link to issue, e.g. `Resolves #NNN`
 - [X] Tests added
 - [X] Branch rebased on top of current master (`git pull --rebase origin master`)
 - [X] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [x] Build is green in Travis CI
 - [X] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/gnatsd/blob/master/LICENSE)

/cc @nats-io/core

Signed-off-by: Waldemar Quevedo <wally@synadia.com>

### CODE REVIEW

File: server/leafnode.go [hunk 1]
```diff
@@ -628,7 +628,12 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
 
 		c.mu.Unlock()
 		// Error will be handled below, so ignore here.
-		c.parse([]byte(info))
+		err = c.parse([]byte(info))
+		if err != nil {
+			c.Debugf("Error reading remote leafnode's INFO: %s", err)
+			c.closeConnection(ReadError)
```
Comment (kozlovic) [inline server/leafnode.go]: I would check for the error returned by parse and just deal with the wrong port error. I would then use `WrongPort` closed reason which may be more appropriate.

**Comment by wallyqs** [comment]
TL;DR; this adds a check so that a leafnode remote connection can detect it is attempting to connect to the client port by noticing whether there are any `connect_url` entries. 

```sh
telnet connect.ngs.global 4222
INFO {..., "connect_urls":["44.228.141.181:4222","35.166.100.73:4222","54.202.186.240:4222"]

telnet connect.ngs.global 7422
INFO {..., "leafnode_urls":["44.228.141.181:7422","54.202.186.240:7422","35.166.100.73:7422"]} 
```

This means that in case of setups were advertisements are disabled the check will fail.  There might other alternatives to this, such as adding metadata to the INFO protocol on whether it is attempting to connect to the client or leafnode port for example.

File: server/leafnode.go [hunk 2]
```diff
@@ -785,12 +790,17 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
...
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
 	if c.leaf == nil || c.isClosed() {
-		return
+		return nil
+	}
+
+	// Prevent connecting to client port.
+	if info.ClientConnectURLs != nil {
```
Comment (kozlovic) [inline server/leafnode.go]: It seems that a server will always have `info.LeafNodeURLs` will always have at least the URL of the server you connect to. So adding that check may help us detect both client and route ports.

File: server/leafnode.go [hunk 3]
```diff
@@ -628,7 +628,12 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
 
 		c.mu.Unlock()
 		// Error will be handled below, so ignore here.
-		c.parse([]byte(info))
+		err = c.parse([]byte(info))
```
Comment (wallyqs) [inline server/leafnode.go]: In case of a malformed JSON error from a server we also now report the error with debug mode here and close the connection earlier.

File: server/leafnode.go (hunk 2 above)
Comment (kozlovic) [reply server/leafnode.go]: That being said, by default the route's port will send the INFO with `leafnode_urls` (except if leafnode disable advertise), so if client advertise is disabled but leafnode is not, then we would not know that we connect to the route. We could add a new field on the INFO that says which type of connection that is. But we would need to make sure that if the field is not set, not to consider it is wrong (would mean old server) and then that would be more reliable for an INFO that has this field set.

File: server/leafnode.go (hunk 1 above)
Comment (wallyqs) [reply server/leafnode.go]: Thanks, ok so only close connection when the error is wrong port reason and fall through for other errors for now?

File: server/leafnode.go [hunk 4]
```diff
@@ -628,7 +628,12 @@ func (s *Server) createLeafNode(conn net.Conn, remote *leafNodeCfg) *client {
...
+		if err != nil {
+			c.Debugf("Error reading remote leafnode's INFO: %s", err)
```
[the remark this answers is not in the document]
Comment (wallyqs) [reply server/leafnode.go]: Agree, I was thinking about that as right now unless using debug it may look as if it was connected. I used debug to be similar to how this error was handled: https://github.com/nats-io/nats-server/pull/1265/files/ecb5008fe33b5373e95183a6fc5fb68345a4a0eb#diff-51766556c6c02cb21c6ac4f9c3fbe375R641

File: server/leafnode.go (hunk 1 above)
Comment (kozlovic) [reply server/leafnode.go]: Yes, so I would have done something like:
```
if err := c.parse([]byte(info)); err == ErrLeafConnectedToWrongPort {
    c.Error(err.Error())
    c.closeConnection(wrongPort)
    return nil
}
```

File: server/leafnode.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kozlovic) [reply server/leafnode.go]: Should be changed to an error too, IMO.

File: server/leafnode.go (hunk 2 above)
Comment (wallyqs) [reply server/leafnode.go]: That's right, though was not sure on whether to check if `leafnode_urls` were missing since seems there was a special condition that suggested that they could be missing here: https://github.com/nats-io/nats-server/blob/ecb5008fe33b5373e95183a6fc5fb68345a4a0eb/server/leafnode.go#L818
I suppose using `leafnode_urls` is better anyway because as a leafnode you would always advertise your own route? Maybe the both checks together would be a good heuristic?

Comment (kozlovic) [reply server/leafnode.go]: But that's why I am a bit uneasy with what we are checking in the INFO. I think that it may be safer to introduce a new field and obviously that will work only for new servers, but with current approach I would be afraid to reject a valid connection simply because there are conditions (that we missed) where those fields (leafnode_urls/client_urls) are not present.

Comment (kozlovic) [reply server/leafnode.go]: Ok, I think that a LEAF INFO will always have leafnode urls with at least the URL of that server. The ROUTE INFO may or may not have leafnode urls, depending of advertise and if that server itself has leafnode or not.
So I think it is safe to assume that an INFO from a LEAF port has always leafnode_urls, so in other words, the absence of it for sure means that it is not a LEAF port. But it is not enough to detect ROUTE port.

Comment (wallyqs) [reply server/leafnode.go]: Yes was not sure that checking for something missing in the INFO fields would work because all the conditions were not clear, so went for checking instead for `connect_urls` being present since that definitely means it is not meant for leafnodes, think that may cover many cases for now without changing the INFO protocol too much... agree maybe the right way would be to have an extra field to detect whether it is a leafnode connection, but also worried about backwards compatibility...

File: server/leafnode_test.go [hunk 5]
```diff
@@ -869,3 +869,66 @@ func TestLeafCloseTLSConnection(t *testing.T) {
...
+
+func (l *captureDebugErrorLogger) Debugf(format string, v ...interface{}) {
+	select {
+	case l.errCh <- fmt.Sprintf(format, v...):
+	default:
+	}
+}
+
+func TestLeafNodeRemoteWrongPort(t *testing.T) {
+	port := 8786
```
Comment (kozlovic) [inline server/leafnode_test.go]: Test could be simplified and use only random port.
Start a server that accepts leafnode connections (no remote defined). Then create an option with a remote leafnode that points to the client port. This can be done this way:
```
u, _ := url.Parse(fmt.Sprintf("nats://%s:%d", oa.Host, oa.Port)
ob.LeafNodes.Remotes = []*RemoteLeafOpts{{URLs:[]*url.URL{u}}}
```
above assume that oa is the options for the server accepting leafnode connections, and ob is for the server creating the leafnode connection to the client port.

**Comment by wallyqs** [comment]
Thanks @kozlovic ! LGTM 💯
